plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.ldapgroups;
+
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.trino.cache.EvictableCacheBuilder;
+import io.trino.spi.classloader.ThreadContextClassLoader;
+import io.trino.spi.security.GroupProvider;
+
+import javax.naming.NamingException;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class LdapGroupsProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupsProvider.class);
+
+    private final LdapHighLevelClient ldapHighLevelClient;
+
+    private final LoadingCache<String, Set<String>> groupListCache;
+
+    @Inject
+    LdapGroupsProvider(LdapGroupsProviderConfig config, LdapHighLevelClient ldapHighLevelClient)
+    {
+        requireNonNull(config, "ldap groups provider config is null");
+
+        this.ldapHighLevelClient = ldapHighLevelClient;
+
+        this.groupListCache = EvictableCacheBuilder.newBuilder()
```

Comment:
Instead of having the caching as a part of `LdapGroupProvider` - can we introduce something like a `CachingGroupProvider` in the `plugin-toolkit` which could wrap on this implementation

Reply:
Created a `CachingGroupProvider` in `plugin-toolkit` with its won `CacheConfig` and module for injecting